akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala

```diff
@@ -247,4 +206,75 @@ private[akka] object AsyncDnsResolver {
     Future.successful(Answer(-1, immutable.Seq.empty[ResourceRecord], immutable.Seq.empty[ResourceRecord]))
 
   case class ResolveFailedException(msg: String) extends Exception(msg)
+
+  // Internal commands to avoid calling private methods in future callbacks
+  private[akka] object Internal {
+    case class Resolve(
+        name: String,
+        requestType: RequestType,
+        resolver: ActorRef,
+        promise: Promise[DnsProtocol.Resolved])
+  }
+
+  // These methods are not in the class so that we can be sure they don't access actor state
+  private def resolveWithSearch(
+      name: String,
+      requestType: RequestType,
+      resolver: ActorRef,
+      settings: DnsSettings,
+      self: ActorRef): Future[DnsProtocol.Resolved] = {
+    if (settings.SearchDomains.nonEmpty) {
+      val nameWithSearch = settings.SearchDomains.map(sd => name + "." + sd)
+      // ndots is a heuristic used to try and work out whether the name passed in is a fully qualified domain name,
+      // or a name relative to one of the search names. The idea is to prevent the cost of doing a lookup that is
+      // obviously not going to resolve. So, if a host has less than ndots dots in it, then we don't try and resolve it,
+      // instead, we go directly to the search domains, or at least that's what the man page for resolv.conf says. In
+      // practice, Linux appears to implement something slightly different, if the name being searched contains less
+      // than ndots dots, then it should be searched last, rather than first. This means if the heuristic wrongly
+      // identifies a domain as being relative to the search domains, it will still be looked up if it doesn't resolve
+      // at any of the search domains, albeit with the latency of having to have done all the searches first.
+      val toResolve = if (name.count(_ == '.') >= settings.NDots) {
+        name :: nameWithSearch
+      } else {
+        nameWithSearch :+ name
+      }
+      resolveFirst(toResolve, requestType, resolver, self)
+    } else {
+      val selfCmd = Internal.Resolve(name, requestType, resolver, Promise())
+      self ! selfCmd
+      selfCmd.promise.future
+    }
+  }
+
+  private def resolveFirst(
+      searchNames: List[String],
+      requestType: RequestType,
+      resolver: ActorRef,
+      self: ActorRef): Future[DnsProtocol.Resolved] =
+    searchNames.headOption match {
+      case Some(searchName) =>
+        val selfCmd = Internal.Resolve(searchName, requestType, resolver, Promise())
+        self ! selfCmd
+
+        if (searchNames.tail.isEmpty) {
+          selfCmd.promise.future
+        } else {
+          selfCmd.promise.future.flatMap { resolved =>
+            if (resolved.records.isEmpty) resolveFirst(searchNames.tail, requestType, resolver, self)
+            else selfCmd.promise.future
+          }(ExecutionContexts.parasitic)
```

Comment:
I don't feel comfortable with this kind of construct. 

We have a `Promise` that will be completed by that same actor when it receive the `Resolve` message. 

And then we use the current thread (ExecutionContexts.parasitic) to flatMap on that Future. 

I maybe be wrong, but that this looks unsafe. But also looks like a more complex way of doing actor programming. If I understand it correctly, we are answering to the sender when a Future is completed after a few iterations with the current actor (sending messages to self until completion). 

Instead, I would pass the sender and all the necessary state back to the actor and let it be completed and answered all together. 

Maybe the original impl was abusing the usage of Future and should have be more Actor oriented. And now we are just continuing on the same direction.

Reply:
My understanding is that the parasitic context would only use the current (aka actor) thread if the future had already completed.  Since the future is only completed after processing a message to self, that won't happen: this code will get run on the thread which completes the future (which would be the thread running the actor in processing that message to self).

That said, I think the cleaner fix may be to separate out the requestId management into an actor that only does that

Reply:
That's an interesting discussion. Worth some experiments (on apart pet project) to understand what will really happen. I'm now curious. :-)

Before we introduce another actor, I think we can experiment with two other paths:

1. move all Future logic to the companion object (as suggested by Patrick) to ensure we never capture Actor state
2. remove the usage of Futures as much as we can nd instead use Messages to self and pipeTo(self) when we can't remove the Future. All state is then passed in the messages.

It might be that the reason that we use so many Futures is to avoid pressure on the mailbox, but that's a wild guess.

Reply:
@leviramsey, you are right about when the parasitic ExecCtx will be used. I didn't check, but simply came to the conclusion that it can't be otherwise. 

The function that we pass to `map`/`flatMap` are only called when there is a result and it's at that moment that it should look for a thread to run it. And it picks then the current thread. 

Anyway, just want to bring this here because this topic was left open. At least open in my head. 😄

Reply:
It does bring up the question of whether it makes sense to have the parasitic context be the default for `pipeTo`, since if that callback isn't one that parasitic was made for, I don't know what is (and I believe the `CompletionStage` piping already basically uses the moral equivalent of parasitic anyway, via `whenComplete` vs. `whenCompleteAsync`).